controllers/storageoscluster/metrics-exporter_operand.go

```diff
@@ -0,0 +1,195 @@
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/kustomize/api/filesys"
+	kustomizetypes "sigs.k8s.io/kustomize/api/types"
+)
+
+const (
+	// kMetricsExporterImageName is the image we target with kustomize
+	// transformation on getMetricsExporterBuilder(). Must match the one
+	// on the manifest
+	kMetricsExporterImageName  = "storageos/metrics-exporter:v0.1.1"
+	metricsExporterImageEnvVar = "RELATED_IMAGE_METRICS_EXPORTER"
+)
+
+type MetricsExporterOperand struct {
+	name            string
+	client          client.Client
+	requires        []string
+	requeueStrategy operand.RequeueStrategy
+	fs              filesys.FileSystem
+	kubectlClient   kubectl.KubectlClient
+
+	prometheusCRDsInstalled bool
+	prometheusCRDsLock      chan struct{}
+}
+
+var _ operand.Operand = &MetricsExporterOperand{}
+
+func (c *MetricsExporterOperand) Name() string                                           { return c.name }
+func (c *MetricsExporterOperand) Requires() []string                                     { return c.requires }
+func (c *MetricsExporterOperand) RequeueStrategy() operand.RequeueStrategy               { return c.requeueStrategy }
+func (c *MetricsExporterOperand) PostReady(ctx context.Context, obj client.Object) error { return nil }
+
+// dependenciesFound dictates wether the dependencies are present
+// and thus we can apply the metrics-exporter resources
+// currently we depend on Prometheus CRDs
+func (c *MetricsExporterOperand) dependenciesFound() bool {
+	return c.prometheusCRDsInstalled
+}
+
+func (c *MetricsExporterOperand) updateDependenciesState(currentState bool) {
+	c.prometheusCRDsLock <- struct{}{}
+	c.prometheusCRDsInstalled = currentState
+	<-c.prometheusCRDsLock
+}
+
+func (c *MetricsExporterOperand) ReadyCheck(ctx context.Context, obj client.Object) (bool, error) {
+	ctx, span, _, log := instrumentation.Start(ctx, "MetricsExporterOperand.ReadyCheck")
+	defer span.End()
+
+	ctx, cancel := context.WithTimeout(ctx, time.Minute)
+	defer cancel()
+
+	if !c.dependenciesFound() {
+		return false, nil
+	}
+
+	nodeDS := &appsv1.DaemonSet{}
+	key := client.ObjectKey{Name: "storageos-node", Namespace: obj.GetNamespace()}
+	if err := c.client.Get(ctx, key, nodeDS); err != nil {
+		return false, err
+	}
+
+	metricsDS := &appsv1.DaemonSet{}
+	key = client.ObjectKey{Name: "storageos-metrics-exporter", Namespace: obj.GetNamespace()}
+	if err := c.client.Get(ctx, key, metricsDS); err != nil {
+		return false, err
+	}
+
+	// metrics-exporter only returns complete metrics when there's one instance running
+	// alongside each node container. Failing to do so means the metrics can be missing
+	// data for some volumes.
+	if metricsDS.Status.NumberReady == nodeDS.Status.DesiredNumberScheduled {
+		log.Info("metrics-exporter is ready", "NumberReady", metricsDS.Status.NumberReady)
+		return true, nil
+	}
+
+	log.Info("metrics-exporter not ready")
```

Comment:
If we add fields for `metricsDS.Status.NumberReady` and `nodeDS.Status.DesiredNumberScheduled` it will be easier to determine why we're not ready

Reply:
Added